### PR TITLE
detect Git submodules properly

### LINKED ISSUE #3666: cannot generate history for Git submodules
Setup with Git submodule described in https://github.com/oracle/opengrok/issues/3351#issuecomment-880625163 leads to NPE in `GitRepository`:
```
14:03:18 WARNING: An error occurred while creating cache for /var/opengrok/src.gitsub/main_repo/submd_repo (GitRepository)
java.lang.NullPointerException
	at org.eclipse.jgit.internal.storage.file.UnpackedObjectCache$Table.index(UnpackedObjectCache.java:115)
	at org.eclipse.jgit.internal.storage.file.UnpackedObjectCache$Table.contains(UnpackedObjectCache.java:76)
	at org.eclipse.jgit.internal.storage.file.UnpackedObjectCache.isUnpacked(UnpackedObjectCache.java:31)
	at org.eclipse.jgit.internal.storage.file.LooseObjects.hasCached(LooseObjects.java:82)
	at org.eclipse.jgit.internal.storage.file.ObjectDirectory.openObjectWithoutRestoring(ObjectDirectory.java:339)
	at org.eclipse.jgit.internal.storage.file.ObjectDirectory.openObject(ObjectDirectory.java:330)
	at org.eclipse.jgit.internal.storage.file.WindowCursor.open(WindowCursor.java:132)
	at org.eclipse.jgit.lib.ObjectReader.open(ObjectReader.java:212)
	at org.eclipse.jgit.revwalk.RevWalk.parseAny(RevWalk.java:917)
	at org.eclipse.jgit.revwalk.RevWalk.parseCommit(RevWalk.java:827)
	at org.opengrok.indexer.history.GitRepository.accept(GitRepository.java:476)
	at org.opengrok.indexer.history.BoundaryChangesets.getBoundaryChangesetIDs(BoundaryChangesets.java:88)
	at org.opengrok.indexer.history.RepositoryWithPerPartesHistory.doCreateCache(RepositoryWithPerPartesHistory.java:85)
	at org.opengrok.indexer.history.Repository.createCache(Repository.java:411)
	at org.opengrok.indexer.history.HistoryGuru.createCache(HistoryGuru.java:556)
	at org.opengrok.indexer.history.HistoryGuru.lambda$createCacheReal$3(HistoryGuru.java:603)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```
This supposedly worked prior to the JGit changes (#1113).

### PR DESCRIPTION
**Pull request by vladak**
This change fixes Git submodule detection. The `SubmoduleWalk` is not really suitable for this and is `RepositoryBuilder()` wrapper anyway.

fixes #3666.

Also I did some semi-related cleanup.

### CODE REVIEW

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java [hunk 1]
```diff
@@ -423,6 +405,7 @@ public boolean fileHasHistory(File file) {
     boolean isRepositoryFor(File file, CommandTimeoutType cmdType) {
         if (file.isDirectory()) {
             File f = new File(file, ".git");
```
Comment (ahornace) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java:407]: Use `Constants.DOT_GIT` here as well?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java [hunk 2]
```diff
@@ -675,8 +659,45 @@ boolean hasFileBasedTags() {
...
     private org.eclipse.jgit.lib.Repository getJGitRepository(String directory) throws IOException {
-        return FileRepositoryBuilder.create(Paths.get(directory, ".git").toFile());
+        File dotGitFile = Paths.get(directory, Constants.DOT_GIT).toFile();
+        if (dotGitFile.isDirectory()) {
+            return FileRepositoryBuilder.create(dotGitFile);
+        }
+
+        // Assume this is a sub-module so dotGitFile is a file.
+        String gitDirValue;
+        if ((gitDirValue = getGitDirValue(dotGitFile)) == null) {
```
Comment (ahornace) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java:689]: Just a matter of taste but I find this cleaner:
```java
String gitDirValue = getGitDirValue(dotGitFile);
if (gitDirValue == null) {
```

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java [hunk 3]
```diff
@@ -675,8 +659,45 @@ boolean hasFileBasedTags() {
         return true;
     }
 
+    /**
+     * @param dotGit {@code .git} file
+     * @return value of the {@code gitdir} property from the file
+     */
+    private String getGitDirValue(File dotGit) {
+        try (Scanner scanner = new Scanner(dotGit, StandardCharsets.UTF_8)) {
```
Comment (ahornace) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java:667]: Nothing wrong with this but maybe the functional approach is more readable?
```java
Files.lines(...)
    .filter(line -> line.startsWith(Constants.GITDIR))
    .map(line -> line.substring(Constants.GITDIR.length()))
    .findFirst()
    .orElse(null);
```

File: opengrok-indexer/src/test/java/org/opengrok/indexer/history/GitRepositoryTest.java [hunk 4]
```diff
@@ -788,4 +793,48 @@ private void removeRecursive(final File cloneRoot) {
...
+        Git newRepo = Git.init().setDirectory(newRepoFile).call();
+        assertNotNull(newRepo);
+
+        // Add this repository as a submodule to the existing Git repository.
+        org.eclipse.jgit.lib.Repository mainRepo = new FileRepositoryBuilder().
+                setGitDir(Paths.get(repository.getSourceRoot(), "git", Constants.DOT_GIT).toFile())
+                .build();
+        try (Git git = new Git(mainRepo)) {
+            git.submoduleAdd().
+                    setURI("file:///" + newRepoFile).
```
Comment (ahornace) [inline opengrok-indexer/src/test/java/org/opengrok/indexer/history/GitRepositoryTest.java:809]: Maybe use `newRepoFile.toPath().toUri().toString()` ?

**Review by ahornace — APPROVED** [review]
Nice job! :)

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java (hunk 2 above)
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java:689]: Yep. Strange how I find this perfectly acceptable in C code and in Java code this looks a bit awkward.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java (hunk 3 above)
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java:667]: depends on personal preference I'd say. To me the classic way looks more readable.

File: opengrok-indexer/src/test/java/org/opengrok/indexer/history/GitRepositoryTest.java (hunk 4 above)
Comment (vladak) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/history/GitRepositoryTest.java]: good idea !

**Comment by vladak** [comment]
The solution for #3315 that would also address this problem however involves more complexity (things like deciding whether to continue repository scan for given directory subtree if there are submodules detected). At this point I'd like to get this fixed rather quickly.

**Pull request merged**